File: src/main/java/raylras/zen/util/ArrayStack.java
package raylras.zen.util;

import java.util.ArrayDeque;
import java.util.Deque;

public class ArrayStack<E> implements Stack<E> {

    private final Deque<E> deque;

    public ArrayStack() {
        this.deque = new ArrayDeque<>();
    }

    @Override
    public void push(E e) {
        deque.push(e);
    }

    @Override
    public E pop() {
        return deque.pop();
    }

    @Override
    public E peek() {
        return deque.peek();
    }

    @Override
    public int size() {
        return deque.size();
    }

    @Override
    public boolean isEmpty() {
        return deque.isEmpty();
    }

    @Override
    public void clear() {
        deque.clear();
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/util/ArrayStack.kt
package raylras.zen.util

class ArrayStack<E> : Stack<E> {
    private val deque = ArrayDeque<E>()

    override fun push(e: E) = deque.addFirst(e)

    override fun pop(): E? = deque.removeFirstOrNull()

    override fun peek(): E? = deque.firstOrNull()

    override fun size(): Int = deque.size

    override fun isEmpty(): Boolean = deque.isEmpty()

    override fun isNotEmpty(): Boolean = deque.isNotEmpty()

    override fun clear() = deque.clear()

    override fun iterator(): Iterator<E> = deque.iterator()
}
